cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -121,6 +123,108 @@ SQLRETURN SQLFreeHandle(SQLSMALLINT type, SQLHANDLE handle) {
 }
 
+SQLRETURN SQLGetDiagFieldW(SQLSMALLINT handleType, SQLHANDLE handle,
+                           SQLSMALLINT recNumber, SQLSMALLINT diagIdentifier,
+                           SQLPOINTER diagInfoPtr, SQLSMALLINT bufferLength,
+                           SQLSMALLINT* stringLengthPtr) {
+  using driver::odbcabstraction::Diagnostics;
+  using ODBC::GetStringAttribute;
+  using ODBC::ODBCConnection;
+  using ODBC::ODBCEnvironment;
+
+  if (!handle) {
+    return SQL_INVALID_HANDLE;
+  }
+
+  if (!diagInfoPtr) {
+    return SQL_ERROR;
+  }
+
+  // Set character type to be Unicode by defualt (not Ansi)
+  bool isUnicode = true;
+  Diagnostics* diagnostics = nullptr;
+
+  switch (handleType) {
+    case SQL_HANDLE_ENV: {
+      ODBCEnvironment* environment = reinterpret_cast<ODBCEnvironment*>(handle);
+      diagnostics = &environment->GetDiagnostics();
+      break;
+    }
+
+    case SQL_HANDLE_DBC: {
+      ODBCConnection* connection = reinterpret_cast<ODBCConnection*>(handle);
+      diagnostics = &connection->GetDiagnostics();
+      break;
+    }
+
+    default:
+      return SQL_ERROR;
+  }
+
+  if (!diagnostics) {
+    return SQL_ERROR;
+  }
+
+  // Retrieve header level diagnostics if Record 0 specified
+  if (recNumber == 0) {
+    switch (diagIdentifier) {
+      case SQL_DIAG_NUMBER: {
+        SQLINTEGER count = static_cast<SQLINTEGER>(diagnostics->GetRecordCount());
+        *static_cast<SQLINTEGER*>(diagInfoPtr) = count;
+        if (stringLengthPtr) {
+          *stringLengthPtr = sizeof(SQLINTEGER);
+        }
+
+        return SQL_SUCCESS;
+      }
+
+      case SQL_DIAG_SERVER_NAME: {
+        const std::string source = diagnostics->GetDataSourceComponent();
+        return GetStringAttribute(isUnicode, source, false, diagInfoPtr, bufferLength,
+                                  stringLengthPtr, *diagnostics);
+      }
+
+      default:
+        return SQL_ERROR;
+    }
+  }
+
+  // Retrieve record level diagnostics from specified 1 based record
+  uint32_t recordIndex = static_cast<uint32_t>(recNumber - 1);
+  if (!diagnostics->HasRecord(recordIndex)) {
+    return SQL_NO_DATA;
+  }
+
+  // Retrieve record field data
+  switch (diagIdentifier) {
+    case SQL_DIAG_MESSAGE_TEXT: {
+      const std::string message = diagnostics->GetMessageText(recordIndex);
+      return GetStringAttribute(isUnicode, message, false, diagInfoPtr, bufferLength,
+                                stringLengthPtr, *diagnostics);
+    }
+
+    case SQL_DIAG_NATIVE: {
+      *static_cast<SQLINTEGER*>(diagInfoPtr) = diagnostics->GetNativeError(recordIndex);
+      if (stringLengthPtr) {
+        *stringLengthPtr = sizeof(SQLINTEGER);
+      }
+
+      return SQL_SUCCESS;
+    }
+
+    case SQL_DIAG_SQLSTATE: {
+      const std::string state = diagnostics->GetSQLState(recordIndex);
+      return GetStringAttribute(isUnicode, state, false, diagInfoPtr, bufferLength,
+                                stringLengthPtr, *diagnostics);
+    }
+
+    default:
```

Comment:
It might be a problem to throw an error for unrecognized, but standard diag types. (Like it breaks apps). Perhaps we just return dummy values for any valid, but not supported fields. For non-standard fields we can report error.

The spec says we should only return SQL_ERROR for when "The DiagIdentifier argument was not one of the valid values."